lib/persistent-knex-model-mixin.js

```diff
@@ -0,0 +1,123 @@
+  }
+
+  Model.prototype.save = function * (options) {
+    const myKnex = knexOrTransaction(knex, options)
+    return myKnex.update(this.getDataPersistent()).into(Model.tableName).where('id', this.id).then()
+  }
+
+  // Delete row based on where options
+  Model.prototype.destroy = function (selectOptions) {
+    return knex(Model.tableName).where(selectOptions).del()
+  }
+
+  Model.fromDatabaseModel = function (data) {
+    if (!data) return null
+    // If result is an array (return from knex select query), take the first object
+    if (Array.isArray(data)) {
+      data = data[0]
+    }
+    const model = new Model()
+    model.setDataPersistent(data)
+    return model
+  }
+
+  Model.fromDataPersistent = function (data) {
+    const model = new Model()
+    model.setDataPersistent(data)
+    return model
+  }
+
+  Model.findAll = function * findAll () {
+    // this returns an array of JSON objects
+    return knex(Model.tableName).select().then()
+  }
+  Model.findById = function * findById (id, options) {
+    const myKnex = knexOrTransaction(knex, options)
+    return myKnex.from(Model.tableName).select().where('id', id).then(fromDatabaseModel)
+  }
+
+  // Optional: transaction
+  // set this to knex transaction if you want this operation to be done inside transaction.
+  Model.create = function * create (values, options) {
+    const model = Model.fromData(values)
+    const modelPersistent = model.getDataPersistent()
+    const myKnex = knexOrTransaction(knex, options)
+    return myKnex.insert(cleanUndefinedAttributes(modelPersistent)).into(Model.tableName)
+  }
+
+  Model.bulkCreate = function * bulkCreate (records, options) {
+    const models = records
+      .map(Model.fromData.bind(Model))
+      .map(_.method('getDataPersistent'))
+
+    yield bulkCreateInternal(models)
+  }
+
+  Model.bulkCreateExternal = function * bulkCreateExternal (records, options) {
+    const models = records
+      .map(Model.fromDataExternal.bind(Model))
+      .map(_.method('getDataPersistent'))
+
+    yield bulkCreateInternal(models)
+  }
+
+  // Helper functions
+  function * bulkCreateInternal (models) {
+    if (models.length > 0) {
+      for (const i in models) {
+        models[i] = cleanUndefinedAttributes(models[i])
+      }
+      return knex(Model.tableName).insert(models)
+    }
+  }
+
+  function knexOrTransaction (knex, options) {
+    return !options ? knex : (!options.transaction ? knex : options.transaction)
+  }
+
+  function fromDatabaseModel (result) {
+    return result && Array.isArray(result) && result.length > 0
+      ? Model.fromDatabaseModel(result) : null
+  }
+
+  // If you update or insert into Oracle an object with undefined attributes,
+  // Oracle would give you an error.
+  // So clean up the undefiend objects.
+  // Input: JSON object
+  // Output: Same JSON object, but undefined attributes are deleted.
+  function cleanUndefinedAttributes (obj) {
+    for (const i in obj) {
```

Comment:
Using lodash you could do `_.omit(obj, _.isNil)`, which also makes it non-mutating